### PR TITLE
Exception transparency in handler/callback interfaces

### PR DESCRIPTION
**Pull request by qualidafial**
Addresses issue #290 

To do:

* [x] ~~Remove `Transaction` callback interface in favor of `TransactionCallback`~~ Rename it to `TransactionalCallback` and move it to sqlobject where it belongs
* [x] Add `TransactionalConsumer` and friend `Transactional.useTransaction(TransactionalConsumer)`
* ~~See if `TransactionFailedException` can be removed~~ nah

### CODE REVIEW

File: core/src/main/java/org/jdbi/v3/tweak/transactions/LocalTransactionHandler.java [hunk 1]
```diff
@@ -165,36 +165,28 @@ public boolean isInTransaction(Handle handle)
...
         catch (Exception e) {
             try {
                 handle.rollback();
             } catch (Exception rollback) {
                 e.addSuppressed(rollback);
             }
-            throw new TransactionFailedException("Transaction failed do to exception being thrown " +
-                                                 "from within the callback. See cause " +
-                                                 "for the original exception.", e);
+            throw (X) e;
```
Comment (stevenschlansker) [inline core/src/main/java/org/jdbi/v3/tweak/transactions/LocalTransactionHandler.java]: Ah, what a sneaky throw ;)

**Comment by stevenschlansker** [comment]
Nice!  Is this ready to merge, or still pending the TODOs?  I think the current state is good, happy to merge as is and handle the rest as followups, or block it -- your call.  Thanks!

File: core/src/main/java/org/jdbi/v3/tweak/transactions/LocalTransactionHandler.java (hunk 1 above)
Comment (qualidafial) [inline core/src/main/java/org/jdbi/v3/tweak/transactions/LocalTransactionHandler.java]: My reasoning is that the try can only throw a `RuntimeException` or `X`. So the cast is safe from a runtime perspective. At compile time `X` is erased to just `Exception`.

Plus there's the whole part about how checked exceptions are a purely compile time construct--the runtime makes no distinction and any method can legally throw any Throwable they like.

**Comment by qualidafial** [comment]
I want to hold this one and merge after the extensions, since the merge effort (and pain) should be less on this PR than on the other one.

File: core/src/main/java/org/jdbi/v3/tweak/transactions/LocalTransactionHandler.java (hunk 1 above)
Comment (stevenschlansker) [inline core/src/main/java/org/jdbi/v3/tweak/transactions/LocalTransactionHandler.java]: Yeah, got it.  I like that it removes one of the catch clauses!

**Comment by qualidafial** [comment]
@stevenschlansker I made one more change: renamed `Transaction` interface to `TransactionalCallback`, moved it from core to sqlobject, fixed its generic signature. Also added `TransactionalConsumer` interface to go with it.

If you're good with these changes, I think this is ready to merge.

**Pull request merged**